File: backend/internal/models/reading_progress.go
package models

import "time"

type ReadingStatus string

const (
	ReadingStatusUnread     ReadingStatus = "unread"
	ReadingStatusInProgress ReadingStatus = "in_progress"
	ReadingStatusFinished   ReadingStatus = "finished"
)

type ReadingProgress struct {
	ID              uint          `gorm:"primaryKey" json:"id"`
	UserID          uint          `gorm:"not null;uniqueIndex:idx_user_book_progress" json:"userId"`
	BookID          uint          `gorm:"not null;uniqueIndex:idx_user_book_progress" json:"bookId"`
	Book            Book          `gorm:"foreignKey:BookID" json:"book,omitempty"`
	Status          ReadingStatus `gorm:"size:20;not null;default:unread;index" json:"status"`
	CurrentPage     int           `gorm:"not null;default:0" json:"currentPage"`
	ProgressPercent float64       `gorm:"not null;default:0" json:"progressPercent"`
	IsDownloaded    bool          `gorm:"not null;default:false;index" json:"isDownloaded"`
	LastReadAt      *time.Time    `json:"lastReadAt,omitempty"`
	CreatedAt       time.Time     `json:"createdAt"`
	UpdatedAt       time.Time     `json:"updatedAt"`
}

File: mobile/lib/core/models/reading_progress.dart
import 'book.dart';

class ReadingProgress {
  const ReadingProgress({
    required this.bookId,
    required this.status,
    required this.currentPage,
    required this.progressPercent,
    required this.isDownloaded,
    this.book,
  });

  final int bookId;
  final String status;
  final int currentPage;
  final double progressPercent;
  final bool isDownloaded;
  final Book? book;

  factory ReadingProgress.fromJson(Map<String, dynamic> json) =>
      ReadingProgress(
        bookId: json['bookId'] as int,
        status: json['status'] as String,
        currentPage: json['currentPage'] as int? ?? 0,
        progressPercent: (json['progressPercent'] as num? ?? 0).toDouble(),
        isDownloaded: json['isDownloaded'] as bool? ?? false,
        book: json['book'] is Map<String, dynamic>
            ? Book.fromJson(json['book'] as Map<String, dynamic>)
            : null,
      );
}
